### PR TITLE
BUG: Allow nifti load to fail

### PR DESCRIPTION
**Pull request by hjmjohnson**
Ensure that attempts to write to invalid file paths causes error to be reported. Ref #1958.

## PR Checklist
- [X] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [X] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)
- [X] Added test (or behavior not changed)

### CODE REVIEW

File: Modules/IO/ImageBase/include/itkIOTestHelper.h [hunk 1]
```diff
@@ -75,55 +75,60 @@ class IOTestHelper
...
-    {
-      std::cerr << "Exception Object caught: " << std::endl << err << std::endl;
-      throw;
+    using WriterType = itk::ImageFileWriter<ImageType>;
+    { // Test valid filename writing
+      auto writer = WriterType::New();
+      writer->SetImageIO(imageio);
+      writer->SetFileName(filename);
+      writer->SetInput(image);
+      try
```
Comment (dzenanz) [inline Modules/IO/ImageBase/include/itkIOTestHelper.h:93]: Why not replace by `ITK_TRY_EXPECT_NO_EXCEPTION` from https://github.com/InsightSoftwareConsortium/ITK/blob/v5.3rc04/Modules/Core/TestKernel/include/itkTestingMacros.h#L130?

File: Modules/IO/ImageBase/include/itkIOTestHelper.h [hunk 2]
```diff
@@ -75,55 +75,60 @@ class IOTestHelper
...
-    {
-      // This is the correct behavior
-      std::cout << "Correctly caught exception for attempting to write to an invalid file." << std::endl;
-      return;
-    }
-    catch (...)
-    {
-      itkGenericExceptionMacro(<< "IO library exception not converted to an itk::ExceptionObject.");
+      writer->SetInput(image);
+      try
```
Comment (dzenanz) [inline Modules/IO/ImageBase/include/itkIOTestHelper.h:116]: `ITK_TRY_EXPECT_EXCEPTION` does not have the same effect here.

**Review by dzenanz — APPROVED** [review]

**Comment by hjmjohnson** [comment]
@dzenanz I tried to use ITK macros for the excepion handling in the tests, but did not have time to figure out why the builds failed due to missing header files.

**Review by dzenanz — APPROVED** [review]

**Comment by jhlegarreta** [comment]
Not sure whether this caused the memory leaks that appeared today:
https://open.cdash.org/viewDynamicAnalysis.php?buildid=8091249

The ones for `itkMeshTest` and `itkFreeSurferMeshIOTestBinary` were already there, but not the NIfTI-related ones. Not sure if the `MINCImageIO` one is related to this either.

**Comment by jhlegarreta** [comment]
> Not sure whether this caused the memory leaks that appeared today:
https://open.cdash.org/viewDynamicAnalysis.php?buildid=8091249

> The ones for itkMeshTest and itkFreeSurferMeshIOTestBinary were already there, but not the NIfTI-related ones. Not sure if the MINCImageIO one is related to this either.

Fix proposed in PR #3545.

**Comment by jhlegarreta** [comment]
The MINC-related Valgrind defects persist:
https://open.cdash.org/viewDynamicAnalysis.php?buildid=8097352

Detail
https://open.cdash.org/viewDynamicAnalysisFile.php?id=9661400

Looks like the changes to the `itkIOTestHelper.h` file in https://github.com/InsightSoftwareConsortium/ITK/pull/3529/files#diff-0e320d0a08552a8dd7eea3ec3fe6ee5f2f6f91208ce9509f51a7a2383fdd34ef could be the source.

The most prior recent changes to MINC classes 
https://github.com/InsightSoftwareConsortium/ITK/commit/d90e635bc07228544c8f26346052969323edac27#diff-5df59ff7e425b43eb59c43686c0a58f24c35cb222328b5f332b30565ccb4d45f

did not incur Valgrind defects:
https://open.cdash.org/viewDynamicAnalysis.php?buildid=8085647

Have not investigated further.

**Comment by issakomi** [comment]
The problems are in MINC IO. The changes in this PR just made the problems visible.

There are several leaks if MINC IO exits with exception, it doesn't clean many allocation.

The easiest one is e.g. at [line 1166](https://github.com/InsightSoftwareConsortium/ITK/blob/678b97eaebed1a1fddcb77f64cfb5e293700edea/Modules/IO/MINC/src/itkMINCImageIO.cxx#L1166)
`if (minew_volume_props(&hprops) < 0)`

That `hprops` is cleaned at the end of _WriteImageInformation()_, [line 1300](https://github.com/InsightSoftwareConsortium/ITK/blob/678b97eaebed1a1fddcb77f64cfb5e293700edea/Modules/IO/MINC/src/itkMINCImageIO.cxx#L1300)
`mifree_volume_props(hprops);`
but there are many exception between the lines and `hprops` is not freed. If  `mifree_volume_props(hprops);` is called before every exception -- this will reduce the number of memory leaks, related to _minew_volume_props_ like this one
```
==15585== 64 bytes in 1 blocks are definitely lost in loss record 374 of 486
==15585==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==15585==    by 0x679D69B: minew_volume_props (volprops.c:45) 
```
But there are many more, unfortunately.

I have done trivial test, without using _itkIOTestHelper.h_, just read MHA and write MNC to wrong path, the leaks are there. If the path is correct -- there are no problems, so this PR has done a good job and discovered ancient problems, IMHO.

**Comment by issakomi** [comment]
This may work
https://github.com/InsightSoftwareConsortium/ITK/compare/master...issakomi:ITK:minc_valgr1
it is rather tricky with `micreate_dimension`, after some point (volume is created) it will be freed by _CloseVolume()_, but before it has to freed with `mifree_dimension_handle`.

**Pull request merged**